packages/sdk/src/cross-chain-messenger.ts

```diff
@@ -2301,15 +2301,47 @@ export class CrossChainMessenger {
       }
 
       if (this.bedrock) {
-        const withdrawal = await this.toLowLevelMessage(resolved, messageIndex)
-        return this.contracts.l1.OptimismPortal.populateTransaction.finalizeWithdrawalTransaction(
+        // get everything we need to finalize
+        const [status, withdrawal] = await Promise.allSettled([
+          this.getMessageStatus(message),
+          this.toLowLevelMessage(resolved, messageIndex),
+        ])
+
+        // handle errors
+        if (status.status === 'rejected' || withdrawal.status === 'rejected') {
+          const errors: Error[] = []
+          for (const res of [status, withdrawal]) {
+            if (res.status === 'rejected') {
+              errors.push(res.reason)
+            }
+          }
+          throw errors.length > 1 ? new AggregateError(errors) : errors[0]
+        }
+
+        // pick the correct finalize method based on if we are relaying or replaying a failed relay
+        const finalizeMethod = (() => {
+          switch (status.value) {
+            case MessageStatus.READY_FOR_RELAY:
+              return this.contracts.l1.OptimismPortal.populateTransaction
+                .finalizeWithdrawalTransaction
+            case MessageStatus.READY_FOR_REPLAY:
+              return this.contracts.l1.L1CrossDomainMessenger
+                .populateTransaction.relayMessage
+            default:
+              throw new Error(
+                `Message is not ready for finalization: ${status.value}`
+              )
+          }
+        })()
+
+        return finalizeMethod(
           [
-            withdrawal.messageNonce,
-            withdrawal.sender,
-            withdrawal.target,
-            withdrawal.value,
-            withdrawal.minGasLimit,
-            withdrawal.message,
+            withdrawal.value.messageNonce,
```

Comment:
Ah doing this won't work. You will need to handle the inputs differently depending on the function. This is fine for `finalizeWithdrawalTransaction` but for `relayMessage` you will need to:

```
        const xdmWithdrawal = this.contracts.l1.L1CrossDomainMessenger.interface.decodeFunctionData(
          'relayMessage',
          withdrawal.value.message
        )

        await finalizeMethod(
          xdmWithdrawal._nonce,
          xdmWithdrawal._sender,
          xdmWithdrawal._target,
          xdmWithdrawal._value,
          xdmWithdrawal._minGasLimit,
          xdmWithdrawal._message
        )
```